src/util/util.ts

```diff
@@ -235,11 +244,20 @@ async function sidecarPost(url: string, tx: string): Promise<any> {
       },
     )
     .then(({ data }) => data)
-    .then(({ cause, data, error, hash }) => {
+    .then(({ cause, data, error, hash, weight, partialFee }) => {
       if (cause || error) {
         throw new Error(`${cause}: ${error} (${data})`);
       }
 
-      return hash;
+      if (hash) {
+        return hash;
+      }
+
+      if (partialFee) {
+        return {
+          weight,
+          partialFee
+        };
+      }
```

Comment:
Two questions about this:

1. It just seems very hacky in general. There must be a more generic way to group return items?
2. The return will either be `{ hash; } ` or `{ weight; class; partialFee; }`. In the latter case, can I group these into a `DispatchInfo` interface? Also, `class` of course is reserved, so how do I pick it out as a field of `DispatchInfo` in line 247?

Reply:
I think the root of your problems goes back to your layers of abstraction. At a glance, it looks like you're trying to do too much in this function (`sidecarPost`). I would suggest creating purpose-built functions, each of which has their own singular return type and each of which calls into the `POST` helper function. You may need to do some additional refactoring to achieve this, but overall I think this is the way in which you want to be thinking :+1:

Reply:
Talked with Joe on offline about investigating a generic `sidecarPost<T>` solution to this problem.

Reply:
+1 with Dan, I would create:
- `sidecarPost<T>`, whose goal is only to send a POST request to the sidecar
- and maybe in `submitTransaction` and `getFeeEstimate`, do the `if (hash) else ...` checks.

>  In the latter case, can I group these into a DispatchInfo interface? Also, class of course is reserved, so how do I pick it out as a field of DispatchInfo in line 247?

Yeah, I guess you could:
```diff
+import { RuntimeDispatchInfo } from '@polkadot/types/interfaces';

-export async function getFeeEstimate(sidecarHost: string, encodedTx: string): Promise<any> {
+export async function getFeeEstimate(sidecarHost: string, encodedTx: string): Promise<RuntimeDispatchInfo> {
```